File: mRemoteNG/Controls/TextBox.vb
Imports System.ComponentModel

' Adapted from http://stackoverflow.com/a/3678888/2101395

Namespace Controls
    Public Class TextBox
        Inherits Windows.Forms.TextBox

#Region "Public Properties"
        <Category("Behavior"),
        DefaultValue(False)> _
        Public Property SelectAllOnFocus As Boolean = False
#End Region

#Region "Protected Methods"
        Protected Overrides Sub OnEnter(ByVal e As EventArgs)
            MyBase.OnEnter(e)

            If MouseButtons = MouseButtons.None Then
                SelectAll()
                _focusHandled = True
            End If
        End Sub

        Protected Overrides Sub OnLeave(ByVal e As EventArgs)
            MyBase.OnLeave(e)

            _focusHandled = False
        End Sub

        Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
            MyBase.OnMouseUp(e)

            If Not _focusHandled Then
                If SelectionLength = 0 Then SelectAll()
                _focusHandled = True
            End If
        End Sub
#End Region

#Region "Private Fields"
        Private _focusHandled As Boolean = False
#End Region
    End Class
End Namespace

File: mRemoteV1.ConvertedToC#/Controls/TextBox.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.ComponentModel;

// Adapted from http://stackoverflow.com/a/3678888/2101395

namespace mRemoteNG.Controls
{
	public class TextBox : Windows.Forms.TextBox
	{

		#region "Public Properties"
		[Category("Behavior"), DefaultValue(false)]
		public bool SelectAllOnFocus { get; set; }
		#endregion

		#region "Protected Methods"
		protected override void OnEnter(EventArgs e)
		{
			base.OnEnter(e);

			if (MouseButtons == MouseButtons.None) {
				SelectAll();
				_focusHandled = true;
			}
		}

		protected override void OnLeave(EventArgs e)
		{
			base.OnLeave(e);

			_focusHandled = false;
		}

		protected override void OnMouseUp(MouseEventArgs e)
		{
			base.OnMouseUp(e);

			if (!_focusHandled) {
				if (SelectionLength == 0)
					SelectAll();
				_focusHandled = true;
			}
		}
		#endregion

		#region "Private Fields"
			#endregion
		private bool _focusHandled = false;
	}
}
